wave_lang/kernel/wave/utils/barriers_utils.py

```diff
@@ -0,0 +1,688 @@
+# Copyright 2025 The IREE Authors
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from enum import Enum, IntFlag, auto
+from dataclasses import dataclass
+from collections import defaultdict
+from typing import Any, Dict, List, Optional, Sequence, Union, Set
+
+import torch.fx as fx
+
+from .graph_utils import propagate_loop_carried_vars
+
+from ..._support.tracing import CapturedTrace
+from ...lang.global_symbols import SHARED_ADDRESS_SPACE
+from ...ops.wave_ops import (
+    AtomicOp,
+    CustomOp,
+    GatherToLDS,
+    Read,
+    Write,
+    get_custom,
+    Iterate,
+    Conditional,
+    NestedRegionOp,
+)
+
+
+@dataclass
+class TargetConfig:
+    target: str
+    has_split_barriers: bool = False
+    has_named_barriers: bool = False
+    max_named_barriers: int = 0
+
+    def __post_init__(self):
+        if "gfx12" in self.target:
+            self.has_split_barriers = True
+
+        if "gfx1250" in self.target:
+            self.has_named_barriers = True
+            self.max_named_barriers = 16
+
+
+class BarrierType(IntFlag):
+    """
+    For RAW Write -> Read: guard fill
+    For WAR Read -> Write: guard ready
+    """
+
+    NONE = 0
+    READY = auto()
+    FILL = auto()
+
+
+@dataclass
+class SyncRequirement:
+    """
+    Synchronization requirements in between producers and consumers.
+    """
+
+    resource: Any = None
+    producers: Sequence[Any] = None
+    consumers: Sequence[Any] = None
+    is_loop: bool = False
+    prod_region: fx.Node = None
+    cons_region: fx.Node = None
+    prod_topo_location: int = -1
```

Comment:
Nit: Can we just call this producer_location? Not sure why there is topo in here.